cou/steps/plan.py

```diff
@@ -435,39 +447,84 @@ def _generate_control_plane_plan(
     return [principal_upgrade_plan, subordinate_upgrade_plan]
 
 
-async def _generate_data_plane_plan(
-    target: OpenStackRelease,
-    analysis_result: Analysis,
+def _separate_hypervisors_apps(
+    apps_data_plane: list[OpenStackApplication],
+) -> tuple[list[OpenStackApplication], list[OpenStackApplication]]:
+    """Separate what is considered hypervisors apps from non-hypervisors apps.
+
+    :param apps_data_plane: Applications from data-plane
+    :type apps_data_plane: list[OpenStackApplication]
+    :return: Tuple containing two lists of hypervisors and non-hypervisors apps
+    :rtype: tuple[list[OpenStackApplication], list[OpenStackApplication]]
+    """
+    hypervisor_apps = []
+    non_hypervisors_apps = []
+    for app in apps_data_plane:
+        if app.charm == "ceph-osd" or app.is_subordinate:
+            non_hypervisors_apps.append(app)
+            continue
+
+        hypervisor_apps.append(app)
+    return hypervisor_apps, non_hypervisors_apps
+
+
+async def _generate_hypervisors_plan(
     args: CLIargs,
-) -> list[UpgradePlan]:  # pragma: no cover
-    """Generate upgrade plan for data plane.
+    analysis_result: Analysis,
+    target: OpenStackRelease,
+    hypervisor_apps: list[OpenStackApplication],
```

Comment:
not-blocking: Since the name of function already contain the "hypervisors", is really important to use `hypervisor_apps` instead of `apps`? This naming seems redundant to me.